torchtune/utils/collate.py

```diff
@@ -133,3 +129,68 @@ def padded_collate_dpo(
     )
 
     return concatenated_input_ids, concatenated_labels
+
+
+def _padded_collate_packed(
+    sample: Dict[str, Any],
+    max_seq_len: int,
+    padding_idx: int = 0,
+    ignore_idx: int = CROSS_ENTROPY_IGNORE_IDX,
+) -> Dict[str, torch.Tensor]:
+    """Pad a batch of packed sequences to max sequence length in the batch, and
+    convert integer lists to tensors. Account for attention mask and position
+    ids.
+
+    This is a sample-wise collator and should not be used with the dataloader.
+
+    Sample should look like::
+        {
+            "tokens": List[int],
+            "labels": List[int],
+            "mask": Tensor,
+            "input_pos": List[int],
+        }
+
+    Args:
+        sample (Dict[str, Any]): A dictionary containing tokens, labels, mask,
+            and position ids
+        padding_idx (int): Padding index for input ids. Defaults to 0.
+        ignore_idx (int): Padding index for labels. Defaults to -100.
+
+    Returns:
+        Collated tokens, labels, mask, and position ids.
+    """
+    tokens = sample["tokens"]
+    labels = sample["labels"]
+    mask = sample["mask"]
+    input_pos = sample["input_pos"]
+
+    # Pad to max sequence length
+    tokens = F.pad(
+        torch.tensor(tokens), (0, max_seq_len - len(tokens)), value=padding_idx
+    )
+    labels = F.pad(
+        torch.tensor(labels), (0, max_seq_len - len(labels)), value=ignore_idx
+    )
+    # For attention mask, simply use identity matrix for the pad tokens
+    mask_pad = torch.eye(max_seq_len - mask.shape[0], dtype=torch.bool)
+    mask = torch.block_diag(mask, mask_pad)
+    # For position ids, continue to increment for pad tokens
+    input_pos_pad = torch.arange(
+        input_pos[-1] + 1, max_seq_len - len(input_pos) + input_pos[-1] + 1
```

Comment:
nit: a bit confusing, maybe just define `next_pos=input_pos[-1] + 1` in a separate line to make the arange clearer